Experimental Announced Component

natalieethell opened the pull request:
#### Pull request checklist

- [X] Addresses an existing issue: https://github.com/OfficeDev/office-ui-fabric-react/issues/5615
- [X] Include a change request file using `$ npm run change`

### Description of changes

#### tl;dr: These changes add a prototype of the new Announced component, with examples, to the experiments package. The majority of my time was spent on making up the examples for the demo page, which I based off of the scenarios we detailed in the initial documentation for this component [here](https://github.com/natalieethell/office-ui-fabric-react/blob/21a1a20ad6589722510170af46f4f67b0aebf146/packages/experiments/src/components/Announced/Announced.md), with some small tweaks.

#### Examples

This PR has **three examples** on the new Announced demo page. Below I have screenshots and screen captures of the example and the Narrator Buddy output.

1. Announced component used for asynchronous resolution of contacts or search results

![image](https://user-images.githubusercontent.com/14134000/47244183-6c9d9300-d3a9-11e8-8ddc-067b715326c7.png)
![image](https://user-images.githubusercontent.com/14134000/47244436-547a4380-d3aa-11e8-9d6a-42e50a328b2c.png)

2. Announced component used for asynchronous page/page content loading

![announcedloadphotos](https://user-images.githubusercontent.com/14134000/47244303-d158ed80-d3a9-11e8-97c5-4ce96ecd8dcf.gif)
![image](https://user-images.githubusercontent.com/14134000/47244514-90ada400-d3aa-11e8-9891-c490904acddd.png)


3. Announced component used for bulk async long running operations

![announcedmoveitems](https://user-images.githubusercontent.com/14134000/47244355-12510200-d3aa-11e8-983e-5ac6015501d8.gif)
![image](https://user-images.githubusercontent.com/14134000/47244679-3b25c700-d3ab-11e8-974f-860cbbf6c837.png)


Ideally, I wanted to make another example that includes the first "Editing text/deletion" scenario listed in the documentation, however I came across some issues, so skipped it for now. *See **Future Work** section below.

#### Initial documentation and "Announced Manager"

Additionally, I updated the documentation based off the discussion had between the initial PR (https://github.com/OfficeDev/office-ui-fabric-react/pull/6305) and now. This includes eliminating all props for the component except for `id` and `message`. We figured that the component may not need a `delay` prop. A feasible option we landed on is to have an external "announced manager" that tracks the number of mails deleted, for example, and debounces them, meaning that X seconds must pass, and then, conditionally renders the Announced component. If we tell developers to build these state machines on their own, all they need is to conditionally render the component with the right text.

In the examples added to the demo page in this PR, I don't have a formal "announced manager" yet. Rather, I am just using unique class methods for each example to conditionally render the announced component based on the scenario. I thought it would be best to get feedback and come to an agreement on these before refining them to another "announced manager" component―I look forward to getting to this point!

#### Feedback

I would appreciate any and all feedback on this component. It is very much so still in its "experimental" phase, but I think it's moving in the right direction. With more eyes and feedback, we can make this component better and better for the components in OUFR that are lacking core accessibility functionality and others out there. Thank you for taking the time to look it over!

#### Future Work

\* To build the "Editing text/deletion" scenario, I tried using DetailsList and I wasn't able to set a specific keyboard event handler on DetailsList for the "delete" key. I tried, instead, placing a contextual menu with "delete" and "rename" buttons into every cell of one column by using onCustomRenderItemColumn, 

![image](https://user-images.githubusercontent.com/14134000/47244034-eaad6a00-d3a8-11e8-9dfd-acc99b817c77.png)

however, passing in an onClick for each button in the menu seemed to break the cells (they fail to even render). If anyone has any thoughts on how we can accomplish this, or any other ideas for the example, let me know. 

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/6778)

KevinTCoughlin commented on packages/experiments/src/components/Announced/examples/Announced.Asynchronous.Example.tsx:
```diff
@@ -0,0 +1,151 @@
+
+    return (
+      <VerticalStack gap={10}>
+        <div>
```
nit: I'd consider making these `<p>` tags rather than `<div>` for semantic reasons unless it involves layout changes.

KevinTCoughlin commented on packages/experiments/src/components/Announced/examples/Announced.Asynchronous.Example.tsx:
```diff
@@ -0,0 +1,151 @@
+        url: `http://placehold.it/100x100`,
+        width: 100,
+        height: 100
+      };
```
nit: You may consider making the `width` and `height` constants and string interpolate them in the `url` value to keep them in sync better.

```ts
    const width: number = 100;
    const height: number = 100;

    return {
        url: `http://placehold.it/${width}x${height}`,
        width,
        height
    };
```

KevinTCoughlin commented on packages/experiments/src/components/Announced/Announced.types.ts:
```diff
@@ -0,0 +1,55 @@
+  /**
+   * The status message provided as screen reader output
+   */
+  message?: string;
```
Would it ever be useful for the `message` to be invoked as a function which returns a `string` from a developer standpoint?

KevinTCoughlin commented on packages/experiments/src/components/Announced/examples/Announced.SearchResults.Example.tsx:
```diff
@@ -0,0 +1,121 @@
+
+  public render(): JSX.Element {
+    return (
+      <div>
```
nit: you can return a [React.Fragment](https://reactjs.org/docs/fragments.html) here `<></>` or `<React.Fragment></React.Fragment>` since these are simply wrapper DIVs.

KevinTCoughlin commented on packages/experiments/src/components/Announced/docs/AnnouncedOverview.md:
```diff
@@ -19,22 +20,22 @@ Editing text, deletion
 1. User renames a file
 
-    When the file is renamed successfully, use the Announced component, coupled with a key, to announce the message immediately. Providing a key will ensure that the announcement is re-read.
+    When the file is renamed successfully, use the Announced component, coupled with a id, to announce the message immediately. Providing a id will ensure that the announcement is re-read.
```
Hmm.. how do I use that new edit in-line feature of Github

natalieethell replied:
Whoops, I must've missed these when I changed `key` to `id`. Nice catch!

betrue-final-final commented:
This is sounding good. I didn't actually get the "photo loading" when I selected one.
Also, does it need to say "Status" every time?

natalieethell commented on packages/experiments/src/components/Announced/docs/AnnouncedOverview.md:
```diff
@@ -1,8 +1,9 @@
-# Announced Component
+The Announced component aims to fill several of the accessibility gaps that exist in various web application experiences.
+It provides text for the screen reader in certain scenarios that are lacking comprehensive updates, particularly those showing
+the completion status or progress of operation(s).
```
I put it on a newline because of line length restrictions :/

betrue-final-final commented:
That's a good question. @jamesedjac?

joschect commented on packages/experiments/src/components/Announced/Announced.base.tsx:
```diff
@@ -0,0 +1,25 @@
+    });
+
+    return (
+      <div role="status" aria-live="assertive" id={id}>
```
I think that we should allow the `aria-live` to be overridden and that it should just default to assertive.

joschect commented on packages/experiments/src/components/Announced/examples/Announced.Asynchronous.Example.tsx:
```diff
@@ -0,0 +1,153 @@
+        if (this.state.seconds % this.delay === 0) {
+          this.setState({
+            announced: (
+              <Announced message={`${this.state.total}/${this.state.photos.length} photos loaded`} id={'announced-' + this.state.total} />
```
Can't the delayed renderer used in the announcer basically have this behavior already?

joschect replied:
I took another look, it can't but it will only delay the first time the component is mounted.

natalieethell replied:
Hmm I'm not entirely sure if it'd do what we want, given that we want to print out the current number of photos loaded so far, and `this.state.total` (total number of photos loaded so far) is changing while the timer is going. So if we passed in `this.state.total` and just specified a delay to pass into the Announced component, I'm not sure that it would handle cases in which `this.state.total` has changed between the time the render is called and the delay. Does that make sense...?

joschect replied:
I think so, sounds right.

natalieethell commented:
Fixing some rebase issues.

natalieethell commented:
@betrue-final-final I think I fixed the "Photo loading" issue. Let me know if you're still not hearing it.

cliffkoh commented on packages/experiments/src/components/Announced/AnnouncedPage.tsx:
```diff
@@ -0,0 +1,5 @@
+import * as React from 'react';
+import { DemoPage } from '../../../../../apps/fabric-website-resources/src/components/DemoPage';
+import { AnnouncedPageProps } from './Announced.doc';
```
Bad import path

cliffkoh commented:
Looks good other than the 1 minor issue I could find in code. I need to checkout your branch and try it out. Will sign off after I've done that!

cliffkoh commented:
There is something weird going on in this PR - looks like CI is encountering the issue too. I can't build experiments - for some reason barely any files appear in the lib folder.

natalieethell commented:
Updated the AnnouncedPage to hopefully remove the build errors. 

I do have one more thing I'd like to mention regarding **re-announcing status messages**.

In the initial PR to add the first draft of the Announced, I mentioned that we discussed using `id` instead of `key` on the component, as `key` is reserved/special in React. I did notice, however, that if the status message in the Announced component stays the same, and only the `id` changes, the screen reader won't re-announce the message. When `key` is used and changed, however, the screen reader will re-announce the status message. We would want this behavior if, for example, someone focused on two sections that were still loading one after another and got the same message "Section loading" for both.

I think it's okay to leave as is in this PR for now, but as we iterate off of this version and discuss other scenarios for the component moving forward, I think it's good to think about what would be better to use instead of `id` to assure that the screen reader reads out all the status messages we send it. One option is removing the Announced component from the DOM after N seconds and re-rendering it, however, that would require agreeing on an N, and we would want to make sure that it's enough time for the screen reader to detect and read out the message.

KevinTCoughlin commented:
I was able to test this in Edge using Narrator and overall this worked great 🎉! The dynamic announcements are fantastic to hear. I think it provides crucial context that would otherwise be lost, but will defer to experts that use screen reader technology daily.

There were some minor issues that I call-out below. @natalieethell reminded me that they are related to the lack of _key_ or a comparable solution which causes React to re-render the DOM containing status. I think we should resolve that before using this in production. Thankfully, a solution there _should_ resolve the issues I encountered.

## Test Scenarios

### Search Results w/ Tag Picker

Was nearly flawless in my testing, however @natalieethell noted in-person that this is subject to the lack of _key_ issue. The scenario affected is if you were to type "b" then "bl", both of which resulted in 2 items retrieved. The status announcement would only occur once outside of the timer.

#### Testing

My steps in testing this scenario:

1. Focus `SearchBox`
1. Type "r"
1. Press `DOWN` to select 2nd option (2 of 2).

#### Findings

The **one issue** I noticed was that selections seem to be announced twice. However, I'm unsure if that is related to this PR.

<pre>
Office UI Fabric React Examples - Microsoft Edge window, Office UI Fabric React Examples, Tag Picker, editable combo box,
<say-as interpret-as="spell">r</say-as>
Caps Lock plus Down Arrow to change to Suggestions view
status, 2 Color Tags Found
<b>
rose, <say-as interpret-as="spell">2</say-as> of <say-as interpret-as="spell">2</say-as>, selected, 
rose, <say-as interpret-as="spell">2</say-as> of <say-as interpret-as="spell">2</say-as>, selected, 
red, <say-as interpret-as="spell">1</say-as> of <say-as interpret-as="spell">2</say-as>, selected, 
red, <say-as interpret-as="spell">1</say-as> of <say-as interpret-as="spell">2</say-as>, selected, 
</b>
Tag Picker, editable combo box,
</pre>

### Async Loading Photos

#### Testing

My steps in testing this scenario:

1. Enable async loading of photos via example toggle
1. Focus container with items that are async loading photos
1. Shift focus using keyboard across photos both loaded and not yet loaded.

#### Findings

I was _surprised_ that `status, Photo loading` was only announced once. Again related to the lack of _key_ causing the element to mutate and therefore the announcement re-read.

```shell
Photo, contains 0 items,
status, Photo loading
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
Photo, contains 0 items,
```

### List Selection Drag & Drop

#### Testing

My steps in testing this scenario:

1. Select the _first_ item in the list.
1. Drag the item _below_ the _second_ item in the list.
1. Drag the item _above_ the _second_ item in the list.

#### Findings

```shell
status, 1 item moved
1 Item 1 1 Roko Kolar 20 MB, selected, 
2 Item 2 2 Roko Kolar 23 MB, non-selected, 
1 Item 1 1 Roko Kolar 20 MB, selected, 
10 Item 10 10 Alex Lundberg 10 MB, non-selected, 
1 Item 1 1 Roko Kolar 20 MB, selected, 
```

I was _surprised_ that the `status, 1 item moved` was only announced _once_ per timeout unless I changed the number of selected items. Again this is due to the _key_ issue.

### Conclusion

The API surface seems reasonable and this component works as expected aside from the issue where `aria-status` isn't re-announced until some time passes or the `status` changes. I'm unsure if _key_ is the _best_ way to resolve this issue, but it does seem like we need to let React know that the element containing `aria-status` should be re-announced. It seems like Narrator decides whether or not to re-read the text based on if that element has changed since last read.

Will wait to approve until we decide whether or not we want to resolve that issue as part of this PR.

Good work @natalieethell and everyone else that pitched in with design and test.

KevinTCoughlin commented on packages/experiments/src/components/Announced/examples/Announced.SearchResults.Example.tsx:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  public componentDidUpdate(prevProps: IAnnouncedSearchResultsExampleProps, prevState: IAnnouncedSearchResultsExampleState): void {
+    const suggestionsContainer = document.getElementsByClassName('ms-Suggestions-container');
```
That's a good catch and suggestion. I think there's a more React way to accomplish this.

KevinTCoughlin commented on packages/experiments/src/components/Announced/Announced.base.tsx:
```diff
@@ -0,0 +1,30 @@
+    return (
+      <div role="status" aria-live={ariaLive} id={id}>
+        <DelayedRender>
+          <div className={this._classNames.screenReaderText}>{message}</div>
```
Would it be useful for developers to be able to pass custom HTML to render within `DelayedRender` here? If so, we can consider using [React.Children](https://reactjs.org/docs/react-api.html#reactchildren).

KevinTCoughlin commented on packages/experiments/src/components/Announced/Announced.base.tsx:
```diff
@@ -0,0 +1,30 @@
+    });
+
+    return (
+      <div role="status" aria-live={ariaLive} id={id}>
```
Would it be useful for developers to customize this element type perhaps with DIV as default? Not sure since the content is typically hidden, but worth consideration.

KevinTCoughlin commented on packages/experiments/src/components/Announced/Announced.styles.ts:
```diff
@@ -0,0 +1,9 @@
+
+export const getStyles = (props: IAnnouncedStyleProps): IAnnouncedStyles => {
+  return {
+    root: [],
```
Not sure if `root` is used

KevinTCoughlin commented on packages/experiments/src/components/Announced/Announced.types.ts:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Root html container for Announced.
+     */
+    root?: string;
```
Same re: this `root`

KevinTCoughlin commented on packages/experiments/src/components/Announced/examples/Announced.Asynchronous.Example.tsx:
```diff
@@ -0,0 +1,177 @@
+        this.setState({ total: this.state.total + 1 });
+      } else if (this.state.total === this.state.photos.length && this.state.complete !== true) {
+        this.setState({ complete: true });
+        this.setState({ announced: undefined });
```
Any reason for not combining these `setState` calls?

natalieethell replied:
Nope, no reason! Just an oversight. Good catch.

KevinTCoughlin commented on packages/experiments/src/components/Announced/examples/Announced.BulkLongRunning.Example.tsx:
```diff
@@ -0,0 +1,192 @@
+import * as React from 'react';
```
On that note, we should see if we can improve the accessibility of these examples before moving the component out of experiments. It might still be worth having for testing via mouse for local development.

I imagine some of the examples evolving into their own components if their usage is common enough. The photos example comes to mind.

natalieethell replied:
Good point! I agree that these are not actions that visually impaired users would perform. The idea here was to represent the "Copying, uploading, or moving many items" scenario. I suppose that a keyboard-events-only version could be illustrated with copying and pasting several items. I'll think more about how we could show this in the demo page.

natalieethell commented on packages/experiments/src/components/Announced/examples/Announced.Asynchronous.Example.tsx:
```diff
@@ -0,0 +1,177 @@
+        <Toggle label="Check to start loading photos" onText="Start/Resume" offText="Pause" onChange={this._onToggleChange} />
+        <ProgressIndicator label={percentComplete < 1 ? 'Loading photos' : 'Finished loading photos'} percentComplete={percentComplete} />
+        <FocusZone elementType="ul" className="ms-AnnouncedExamples-photoList">
+          {this._renderAnnounced()}
```
It does not, good catch!
